### ISSUE TITLE
Doesn't support source & output files that are directly contained by the Fedora object

### ISSUE THREAD

**Issue by flyingzumwalt** [projecthydra/hydra-derivatives#60]
This is shown in PR #59

**Comment by awead** [comment]
@flyingzumwalt seems like the solution should be that hydra-derivatives accepts parameters for these. Is the gem currently making too many assumptions about the model?

**Comment by flyingzumwalt** [comment]
@awead one solution would be to use `.send` instead of `.attached_files`, but I bet there is a reason that it didn't already do that? 

### Example

```ruby
def source_file
    object.send(source_name.to_sym)
end
```
instead of
```ruby
def source_file
    object.attached_files[source_name.to_s]
end
```

**Comment by flyingzumwalt** [comment]
or maybe rely on a retrieve_attached_file method


```ruby
def retrieve_attached_file(object, attachment_ref)
    object.send(attachment_ref.to_sym)
end

def source_file
    retrieve_attached_file(object, source_name)
end

def output_file(path)
        # first, check for a defined file
        output_file = if retrieve_attached_file(object, path)
          object.attached_files[path]
        else
          ActiveFedora::File.new("#{object.uri}/#{path}").tap do |file|
            object.attach_file(file, path)
          end
        end
      end
```

**Comment by flyingzumwalt** [comment]
I would love it if hydra-derivatives didn't assume that you were using fedora at all.  Let the `object` know how persistence happens, and let the calls to `transform_file` tell hydra-derivatives which methods to call on that object in order to retrieve the source and return the output.

**Comment by flyingzumwalt** [comment]
Then hydra-derivatives would become the first hydra gem that's  truly useful outside of the hydra/fedora sphere.

**Comment by awead** [comment]
:+1: to that.

**Comment by awead** [comment]
@flyingzumwalt specify a default `retrieve_attached_file` method that is the current process using attached_files, but the implementor overrides that? Or make it explicit:
``` ruby
class Hydra::Derivatives::Processor::Base

  def retrieve_attached_file
    raise NotImplementedError, "You must specify a method for retrieving the file's content"
  end

end
```

**Comment by flyingzumwalt** [comment]
That doesn't work.  That would mean all of your Processors have to choose
which attached_file strategy to assume.  I should be able to use the Image
processor, etc. regardless of how files are attached to my object.

**Comment by awead** [comment]
fair enough!

**Comment by cjcolvar** [comment]
:+1: to this work, hydra-derivatives assumptions about the model keep us from adopting it.

**Comment by flyingzumwalt** [comment]
Solving this by refactoring Processors to use a RetrieveSourceFile service for retrieving source file and a PersistOutputFile service for persisting output file.  These services can be configured in the initializer and/or in your calls to transform_file.

**Comment by kevinreiss** [comment]
See f48c2e55257, 946f581a5b4f7e.

**Comment by tampakis** [comment]
Fixed in #61. Test coverage in #66.
